File: CS50X/pset06/sentimental-credit/credit.py
import sys


def main():
    # Prompt user for a credit card number
    credit_card = input("Number: ")

    # Calculate Luhn’s Algorithm
    check_sum = calculate_check_sum(credit_card)

    # If is not a valid credit card number, exit
    if not is_valid_credit_card(check_sum):
        print("INVALID")
        sys.exit()

    # Print the type of credit card
    print_credit_card(credit_card)


def calculate_check_sum(number):

    # Flipping credit card number
    reversed_number = number[::-1]
    check_sum = 0

    # Multiply every two number by 2, starting with the second to the last
    for digit in reversed_number[1::2]:

        digit = int(digit) * 2

        # If the result is greater than 10, add each digit between them
        if digit > 9:
            digit1 = digit // 10
            digit2 = digit % 10
            check_sum += digit1 + digit2
        # Else, sum digit
        else:
            check_sum += digit

    # Take the result of the previus sum and add the numbers that were not multiplied
    for digit in reversed_number[0::2]:
        check_sum += int(digit)

    return check_sum


def is_valid_credit_card(number):

    # If the second number of the result of Luhn’s Algorithm is 0, is a valid credit card number
    if (number % 10) == 0:
        return True
    # Else, it doesn't
    else:
        return False


def print_credit_card(number):

    length = len(number)

    # Check the lenght and the initial digits of the credit card number and return the type
    if length == 15 and int(number[:2]) in [34, 37]:
        print("AMEX")
    elif length == 16 and int(number[:2]) in [51, 52, 53, 54, 55]:
        print("MASTERCARD")
    elif (length == 13 or length == 16) and int(number[:1]) == 4:
        print("VISA")
    else:
        print("INVALID")


main()

File: CS50X/pset01/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Prompt user for a credit card number
    long number = get_long("Number: ");

    // Check the number length
    long pivot = number;
    int length = 0;

    while (pivot > 0)
    {
        length++;
        pivot = pivot / 10;
    }

    // --- Calculate checksum ---
    // 1. Multiply every two number by 2, starting with the second to the last
    // 2. Sum each product between then
    long checksum = 0;
    long remainder = 1;
    long x = 100;
    long y = 10;
    int c = 0;

    while (c < length)
    {
        pivot = number % x;
        remainder = pivot / y;
        remainder = remainder * 2;

        if (remainder > 9)
        {
            pivot = remainder % 10;
            remainder = remainder / 10;
            checksum = checksum + pivot + remainder;
        }
        else
        {
            checksum = checksum + remainder;
        }
        x = x * 100;
        y = y * 100;
        c++;
    }

    // 3. Take the result of the previuo sum and add then with the numbers that were not multiplied
    // by two
    remainder = 1;
    x = 10;
    y = 1;
    c = 0;

    while (c < length)
    {
        pivot = number % x;
        remainder = pivot / y;
        checksum = checksum + remainder;

        x = x * 100;
        y = y * 100;
        c++;
    }

    // 4. Take the result of the previous sum and check if the second number is equal to '0'
    // 5. if is TRUE, check the credit card type according to the number of digits and their initial
    // digits
    int check;
    int a = checksum % 10;

    if (a == 0)
    {
        if (length == 15)
        {
            check = number / 10000000000000;
            if (check == 34 || check == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (length == 13)
        {
            check = number / 1000000000000;
            if (check == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (length == 16)
        {
            check = number / 1000000000000000;
            if (check == 5)
            {
                check = number / 100000000000000;
                if (check == 51 || check == 52 || check == 53 || check == 54 || check == 55)
                {
                    printf("MASTERCARD\n");
                }
                else
                {
                    printf("INVALID\n");
                }
            }
            else if (check == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
